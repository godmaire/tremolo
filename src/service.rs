use std::collections::HashMap;

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Definition {
    #[serde(default, rename = "service")]
    services: HashMap<String, Service>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Service {
    image: String,
    #[serde(default = "default_port")]
    port: u16,

    #[serde(default)]
    env: HashMap<String, String>,
    #[serde(default)]
    labels: Vec<String>,
}

fn default_port() -> u16 {
    8080
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn basic_example() {
        let input = r#"
            service "test" {
                image = "nginx:latest"
            }

            service "fake" {
                image = "fake:latest"
            }
        "#;

        let res: Definition = hcl::from_str(input).unwrap();
        let res = format!("{res:#?}");
        insta::assert_snapshot!(res);
    }
}
