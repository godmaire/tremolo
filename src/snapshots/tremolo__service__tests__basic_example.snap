---
source: src/service.rs
expression: res
---
Definition {
    services: {
        "fake": Service {
            image: "fake:latest",
            port: 8080,
            env: {},
            labels: [],
        },
        "test": Service {
            image: "nginx:latest",
            port: 8080,
            env: {},
            labels: [],
        },
    },
}
